### PR TITLE
crimson/onode-staged-tree: fix tree_cursor_t::Cursor to be aware of extent duplication

### PR DESCRIPTION
**Pull request by cyx1231st**
Previously, tree_cursor_t::Cache is not aware of the COWed extent, and still point to the read-only version of value header.

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by cyx1231st** [comment]
@athanatos 
Now `unittest-seastore` in wip-seastore-2021-03-04-010022 stops at
`ceph/src/crimson/os/seastore/onode_manager/staged-fltree/node_delta_recorder.h:29: ceph::bufferlist crimson::os::seastore::onode::DeltaRecorder::get_delta(): Assertion '!is_empty()' failed.`
Which I think is normal because I see `TODO: fill in recorder` that recorder is not filled yet in `FLTreeOnode` during mutation.

**Comment by athanatos** [comment]
I'd be ok with merging https://github.com/ceph/ceph/pull/39828 first and adding a test at the FLTreeOnodeManager level like https://github.com/athanatos/ceph/commit/694fa5ece9c9d205d8565c394ae1082f9e1dac3c .  I think we need functional test coverage at that level, however, as the FLTree implementation is simply too complicated to debug from a higher level.

**Comment by athanatos** [comment]
Hmm, https://github.com/athanatos/ceph/commits/sjust/wip-seastore-2021-03-08-124230 with this patch crashes on ./bin/unittest-fltree-onode-manager apparently during cursor destruction (https://gist.github.com/athanatos/cb5582b517f0a34b328320588dc66bb5).

**Comment by cyx1231st** [comment]
@athanatos 

I can't reproduce it locally, `unittest-fltree-onode-manager` still stops at `Assertion '!is_empty()'` in my environment, under debug build and `-DWITH_ASAN=ON`.

However, from your gist I can get some clues. Cursor's cached key-view is still pointing to the read-only version of extent even when the extent is duplicated but layout has not changed yet. It should be good normally because `CachedExtent` holds a reference to the `prior_instance`. But it becomes a problem after the transaction submission when try to destruct the `Cursor`, `LeafNode::do_untrack_cursor` will try to validate the key-view, but the `prior_instance` could be already freed.

**Comment by athanatos** [comment]
Apologies, that branch didn't actually have the recorder fixes committed.  https://github.com/athanatos/ceph/tree/sjust/wip-seastore-2021-03-08-195742 should do the trick.

**Comment by cyx1231st** [comment]
Both `unittest-fltree-onode-manager` and `unittest-seastore` are working now in wip-seastore-2021-03-08-195742, cherry-picking https://github.com/ceph/ceph/pull/39911/commits/c7ab78fad984c1e8f2101747add47ca01e9294c1.

**Comment by cyx1231st** [comment]
@athanatos Could you include your commits "fltree_onode_manager: fill in delta recorder" or maybe also "onode_tree: add FLTreeOnodeManager unittest", so I can extend "test_fltree_onode_manager.cc" base on these?

**Comment by athanatos** [comment]
@cyx1231st I pushed those two commits as part of https://github.com/ceph/ceph/pull/39828 -- though I commented out the portion of the unit test that currently fails.

**Comment by cyx1231st** [comment]
Changeset:
- Cleanup existing test utilities for reuse;
- Implemented FLTreeOnodeManager unittest;
- Fixed a minor issue in FLTreeOnodeManager;

@athanatos please review if the FLTreeOnodeManager unittest is sufficient;

**Review by tchaikov — APPROVED** [review]
LGTM apart from couple nits.

File: src/crimson/os/seastore/onode_manager/staged-fltree/tree_utils.h [hunk 1]
```diff
@@ -361,8 +359,8 @@ class TreeBuilder {
     return seastar::async([this, &t] {
       logger().info("Verifing insertion ...");
       auto iter = kvs.begin();
-      while (!iter.is_end()) {
-        auto p_kv = iter.get_p_kv();
+      while (iter != kvs.end()) {
```
[the remark this answers is not in the document]
Comment (cyx1231st) [reply src/crimson/os/seastore/onode_manager/staged-fltree/tree_utils.h]: fixed

File: src/crimson/os/seastore/onode_manager/staged-fltree/tree_utils.h [hunk 2]
```diff
@@ -371,9 +369,9 @@ class TreeBuilder {
       logger().info("Verifing range query ...");
       iter = kvs.begin();
       auto cursor = tree->begin(t).unsafe_get0();
-      while (!iter.is_end()) {
+      while (iter != kvs.end()) {
```
[the remark this answers is not in the document]
Comment (cyx1231st) [reply src/crimson/os/seastore/onode_manager/staged-fltree/tree_utils.h]: fixed

File: src/test/crimson/seastore/onode_tree/test_fltree_onode_manager.cc [hunk 3]
```diff
@@ -63,73 +83,147 @@ struct fltree_onode_manager_test_t :
...
+    with_transaction([this, &it, f=std::move(f)] (auto& t) {
+      auto p_kv = *it;
+      auto onode = manager->get_or_create_onode(
+          t, p_kv->key).unsafe_get0();
+
+      std::invoke(f, t, *onode, p_kv->value);
+
+      std::vector<OnodeRef> onodes;
+      onodes.emplace_back(onode);
+      manager->write_dirty(t, onodes).unsafe_get0();
```
[the remark this answers is not in the document]
Comment (cyx1231st) [reply src/test/crimson/seastore/onode_tree/test_fltree_onode_manager.cc]: fixed

**Review by athanatos — APPROVED** [review]

**Pull request merged**